Use enum VisitProposalResponse to decide which view controller to display

olivaresf opened the pull request:
In similar fashion to https://github.com/joemasilotti/TurboNavigator/pull/48, sometimes a bit more control is needed when requesting a view controller from `TurboNavigator`'s delegate.

This PR fixes two issues:

First, `controller(_:forProposal:) -> UIViewController?` is implicitly hiding its behavior of allowing or rejecting a visit proposal. It was deciding whether to allow (by returning a VC) or reject (by returning nil) a proposal. I've made that explicit by returning either `.allow...` or `.reject` so that it's clear that you may decide whether a VisitProposal should proceed.

Second, I like that the default view controller is a `VisitableViewController` since Turbo is the default; I've just modified it so that creation of this default controller happens after VisitProposal behavior was decided. This way, if the delegate wants to return a custom controller, we don't need to create a default `VisitableViewController`.

I've also updated the tests and added a new one. I don't think we were checking that the view stack was not modified when the VisitProposal was rejected (previously by returning nil, now by returning `.reject`. 

This is a breaking change, so I don't know how you want to handle it.

olivaresf committed:
Use enum VisitProposalResponse to decide which view controller to display

olivaresf commented on Tests/TurboNavigatorTests.swift:
```diff
@@ -271,6 +284,14 @@ final class TurboNavigatorTests: XCTestCase {
 
 private class EmptyDelegate: TurboNavigationDelegate {
```
I wasn't sure whether to change EmptyDelegate or not 🤔. `MockDelegate` didn't sound right, so I stuck with Empty. Happy to rename it.

joemasilotti replied:
What about `TestNavigationDelegate`? That aligns with other similar classes in the turbo-ios test suite.

olivaresf commented on Tests/TurboNavigatorTests.swift:
```diff
@@ -77,6 +77,19 @@ final class TurboNavigatorTests: XCTestCase {
         assertVisited(url: proposal.url, on: .main)
     }
+    
+    func test_default_default_default_visitProposalResponseCancelsNavigation() {
```
The naming was a bit strange. I feel I need more context as to why there's 3 defaults 😅.

joemasilotti replied:
Ha, agreed. Each option is separated by an underscore. So in the [handled flows table](https://github.com/joemasilotti/TurboNavigator#handled-flows), this corresponds to a default _current context_, a default _given context_, and a _default given presentation_.

So I think your test should actually read `func test_none_visitProposalResponseCancelsNavigation()` since the first two options don't matter. All we care about is the third.

Let's move this to the bottom of the tests, also. To better align with the table.

olivaresf replied:
🤦🏻‍♂️ I should've RTFM. The comment at the top explains it. Thanks, Joe. Moved the test and renamed it.

jayohms commented on Sources/TurboNavigationDelegate.swift:
```diff
@@ -10,9 +10,14 @@ public protocol TurboNavigationDelegate: AnyObject {
+    ///
+    /// - Parameter proposal: navigation destination
+    /// - Returns: how to react to the visit proposal
+    func response(forProposal proposal: VisitProposal) -> VisitProposalResponse
```
I think we should come up with a different name for this. A `response` is language we use for an actual visit/http responses, which would be confusing long-term for the result of a proposal.

olivaresf replied:
I'd like to convey that this isn't about what to display, but about whether to display anything at all given a VisitProposal. What do you think of...

`shouldVisit(_ proposal: VisitProposal) -> VisitDecision`?

jayohms replied:
That's definitely an improvement. Would it make sense to split up the decision whether you should navigate and obtaining the controller instance?

For example, we could have:
```
shouldVisit(_ proposal: VisitProposal): Bool // or a simple enum if it makes sense
```

Then, if the app determines you should continue the visit, call something like:
```
controller(for controllerName: String): UIViewController
```

The `controllerName` would be obtained from the path config `controller-name` value. (Sorry, I know I'm jumping ahead to a future PR that we discussed, but it seemed relevant here.)

olivaresf replied:
Yes. I wanted to keep it in one function, but I think I'm just forcing the issue a bit too much. 

I'll add `shouldVisit(_ proposal: VisitProposal): Bool`, and then change`controller(_, forProposal:)`'s return value from `UIViewController?` to `UIViewController`. Then I can build the `controller-name` on top of this.

jayohms replied:
Cool, that sounds like a great idea to me 👍

joemasilotti replied:
I prefer to keep everything in a single function. It should make it easer to get started with Turbo Native, even if the naming might not be perfect.

What about something like this?

```swift
/// Return from `handle(proposal:)` to route a custom controller.
enum ProposalResult {
    /// Route a `VisitableViewController`
    case accept

    /// Route a custom `UIViewController` or subclass
    case custom(UIViewController)

    /// Cancel navigation and do nothing
    case reject
}

/// Implement to return custom view controllers to route.
/// - Parameter proposal: The clicked link
/// - Returns: A result accepting the default controller or providing a custom one.
func handle(proposal: VisitProposal) -> ProposalResult
```

jayohms replied:
This is a nice improvement and easy to understand 👍 The only downside I see is we then can't codify the `view-controller` path config lookup convention through the API explicitly, but maybe that's ok.

olivaresf replied:
I've used Joe's suggestion, but I've slightly modified case names:

````swift
enum ProposalResult {
    case accept
    case acceptCustom(UIViewController)
    case reject
}
````

I will take a look at the `view-controller` option in a follow-up PR.

jayohms commented on Tests/TurboNavigatorTests.swift:
```diff
@@ -271,6 +284,14 @@ final class TurboNavigatorTests: XCTestCase {
     func session(_ session: Session, didFailRequestForVisitable visitable: Visitable, error: Error) {}
+    
+    func response(forProposal proposal: VisitProposal) -> VisitProposalResponse {
+        if proposal.url.path == "/cancel" {
+            return .reject
+        }
```
Maybe this is part of another PR you have in mind, but we should discourage inspecting the path to make decisions, since multiple path patterns can be tied to a single set of properties — and can be updated remote. Instead, defer to a custom path config property that can be set for any matching patterns.

olivaresf replied:
This is purely for testing that the navigation stack is not modified when the delegate returns `reject`. 

> we should discourage inspecting the path to make decisions, since multiple path patterns can be tied to a single set of properties

This is probably out of scope for this PR. We should discuss it offline.

joemasilotti replied:
Alternatively, we could bundle a path configuration for the tests and use that. Then we wouldn't even need this `TurboNavigationDelegate` subclass because the routing would come for free.

joemasilotti replied:
Regardless, a custom path configuration property to determine controllers inside of Turbo Navigator / turbo-ios sounds great to me.

jayohms replied:
> Alternatively, we could bundle a path configuration for the tests and use that. Then we wouldn't even need this TurboNavigationDelegate subclass because the routing would come for free.

I'm definitely in favor of this 👍. This is what we do in turbo-android: 
- https://github.com/hotwired/turbo-android/blob/main/turbo/src/main/assets/json/test-configuration.json
- https://github.com/hotwired/turbo-android/blob/main/turbo/src/test/kotlin/dev/hotwire/turbo/nav/TurboNavRuleTest.kt#L68

joemasilotti replied:
@olivaresf, I think this is the only outstanding issue.

Do you have appetite for it? If not, I'm happy to take a pass today.

joemasilotti replied:
I added a test path configuration to use for routing `/cancel`.

How do you both feel about routing a `URL` in this test vs. the `VisitProposal` in all the others? I added a comment but want to make sure it isn't making things more confusing.

Once this is resolved this PR is good to merge in!

olivaresf replied:
Leaving it as-is works for me. I don't feel strongly about it either way 👍🏻.

joemasilotti replied:
Taking another look today, we can do this without a path configuration _or_ routing based on URL. We are already customizing the `VisitProposal` directly under test. Let's continue using that pattern for this one as well!

I think that testing the interaction with path configuration is out of scope for this library. That assigns properties to proposals, we shouldn't have to test or verify that at this level of abstraction.

joemasilotti commented:
Thanks for this, @olivaresf! It's exciting to see some fresh eyes on the project as we get close to merging this upstream.

I think I addressed everything inline with comments outside of this being a breaking change. Since it hasn't hit v1.0 I'm fine with releasing a new minor release and noting the breaking changes in the CHANGELOG.

olivaresf committed:
Rename EmptyDelegate -> TestNavigationDelegate

olivaresf committed:
Rename visit proposal cancelation test

olivaresf committed:
Rename VisitProposalResponse -> ProposalResult

olivaresf committed:
Rename response(forProposal:) -> handle(proposal:)

olivaresf commented:
This should be ready for another review. I think I addressed all comments. The test changes (i.e. including the PathConfig) seem to be much bigger than this PR.

joemasilotti committed:
Code formatting

joemasilotti committed:
Move ProposalResult to its own file

joemasilotti committed:
Route /cancel via path configuration in tests

joemasilotti committed:
Use VisitProposal directly

We are already customizing the VisitProposal, no need to bring in a path
configuration for this. It still avoids the initial issue of routing
based on a URL.
